### PR TITLE
Clean Endpoint shutdown

### LINKED ISSUE #4332: Endpoint stop can fail which leaves the running endpoint instance in an intermediate state
For example if [stopping the receivers fails](https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Unicast/RunningEndpointInstance.cs#L45) the feature startup tasks won't be stopped and neither the transport infrastructure. Even more the builder will not be disposed and the running endpoint instance will never set the stopped flag. Any other thread trying to reenter stop would execute the stopping logic again.

In my opinion the stopping the endpoint should be as fail-safe as possible. Should we make sure that we

- Concurrently close the receivers, log any failed stopping of a receiver with a WARN or INFO level
- Stop all feature startup tasks and log failed ones with WARN or INFO level
- Stop the transport infrastructure and log failures with WARN or INFO level
- Clear the settings
- Dispose the builder 
- Make sure we set the stopped flag

Another approach would be to force implementors to not throw. But that is pretty difficult to achieve and you can only try to achieve it with documentation which probably nobody ever reads.

Thoughts?

### PR DESCRIPTION
**Pull request by danielmarbach**
Closes #4332

### CODE REVIEW

File: src/NServiceBus.Core/Features/FeatureStartupTaskController.cs [hunk 1]
```diff
@@ -22,8 +22,19 @@ public Task Start(IBuilder builder, IMessageSession messageSession)
...
         {
-            await instance.PerformStop(messageSession).ConfigureAwait(false);
-            DisposeIfNecessary(instance);
+            try
+            {
+                await instance.PerformStop(messageSession).ConfigureAwait(false);
+            }
+            catch
+            {
+                // TODO: Should we log here?
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Features/FeatureStartupTaskController.cs]: Should we log here? if yes DEBUG level?

File: src/NServiceBus.Core/Unicast/RunningEndpointInstance.cs [hunk 2]
```diff
@@ -31,28 +31,31 @@ public RunningEndpointInstance(SettingsHolder settings, IBuilder builder, List<T
...
                 }
 
                 Log.Info("Initiating shutdown.");
 
+                // Cannot throw by design
                 await StopReceivers().ConfigureAwait(false);
                 await featureRunner.Stop(messageSession).ConfigureAwait(false);
+                // Can throw
                 await transportInfrastructure.Stop().ConfigureAwait(false);
+            }
```
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/RunningEndpointInstance.cs]: Should we add a catch and a log statement here?

**Comment by danielmarbach** [comment]
@Particular/nservicebus-maintainers please review. Also added two questions

File: src/NServiceBus.Core/Features/FeatureStartupTaskController.cs (hunk 1 above)
Comment (andreasohlund) [inline src/NServiceBus.Core/Features/FeatureStartupTaskController.cs]: >Should we log here? i

It seems reasonable to log exceptions during shutdown

>if yes DEBUG level?

Why only DEBUG, isn't this an indication of some issue => log a WARN?

**Comment by danielmarbach** [comment]
@Particular/nservicebus-maintainers updated the logging part

File: src/NServiceBus.Core/Unicast/RunningEndpointInstance.cs (hunk 2 above)
Comment (danielmarbach) [inline src/NServiceBus.Core/Unicast/RunningEndpointInstance.cs]: Done

**Comment by andreasohlund** [comment]
Would there be value in a acceptance test that documents that .stop() swallows exceptions? (use the fake transport to throw on stop)

**Comment by danielmarbach** [comment]
@andreasohlund Pushed the tests. Please review

**Pull request merged**